### PR TITLE
feat(plugin-sitemap): add sitemap plugin (close #337)

### LINKED ISSUE #337: [Feature request] Sitemap Page request
I like the vuepress a lot 😁, but in Seo I found a. Issue, I am using head yaml to add meta tags and schema tags in body as they are allowed in body also but what about sitemap the only thing preventing me to migrate. It will be great If sitemap will be included in the core.

### PR DESCRIPTION
**Pull request by Mister-Hope**
Close #337

### CODE REVIEW

**Comment by Mister-Hope** [comment]
Addtional: CI failed because of coverage, but since this plugin is deeply coupled with vuepress plugin api, it's hard to add some tests.

File: packages/@vuepress/plugin-sitemap/src/node/index.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { generateSiteMap } from './sitemap'
+
+import type { Plugin } from '@vuepress/core'
+import type { SitemapOptions } from './types'
+
+export * from './types'
+
+const sitemapPlugin: Plugin<SitemapOptions> = (options, app) => {
+  const { themeConfig } = app.options
+  const hostname = options.hostname || (themeConfig.hostname as string)
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/src/node/index.ts:12]: We should avoid reading themeConfig in plugins

**Review by meteorlxy — CHANGES_REQUESTED** [review]
Follow the same file structure as other plugins, i.e. only use `node/index.ts` to import/export things, and use a `node/sitemapPlugin.ts` file to declare the plugin

File: docs/reference/plugin/sitemap.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+- Default: `0.5`
+
+页面优先级，范围 0~1
+
+## Plugin options
+
+### hostname
+
+- Type: `string`
+- Required: Yes
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:35]: Required: `true` / `false`

Be consistent with other docs

File: docs/reference/plugin/sitemap.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+- Type: `string[]`
+- Required: No
+
+Pages not included
+
+### outFile
+
+- Type: `string`
+- Default value: `"sitemap.xml"`
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:56]: - Default:

File: packages/@vuepress/plugin-sitemap/package.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    "@vuepress/client": "2.0.0-beta.21",
+    "@vuepress/core": "2.0.0-beta.21",
+    "@vuepress/plugin-git": "2.0.0-beta.21",
+    "@vuepress/shared": "2.0.0-beta.21",
+    "@vuepress/utils": "2.0.0-beta.21",
+    "chalk": "^4.1.1",
+    "fs-extra": "^10.0.0",
+    "sitemap": "^7.0.0"
+  },
+  "devDependencies": {},
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/package.json:42]: remove

File: packages/@vuepress/plugin-sitemap/src/node/generateSitemap.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  priority?: number
+  img?: SitemapImageOption[]
+  video?: SitemapVideoOption[]
+  alternateLinks?: SitemapLinkOption[]
+}
+
+const stripLocalePrefix = (
+  page: Page
+): {
+  /** path of '/' */
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/src/node/generateSitemap.ts:28]: Use `//` for single line comments

File: packages/@vuepress/plugin-sitemap/src/node/types/frontmatter.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import type { SitemapImageOption, SitemapVideoOption } from './sitemap'
+
+export interface SitemapFrontmatterOption {
+  /**
+   * 更新频率
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/src/node/types/frontmatter.ts:5]: Remove zh comments

**Review by meteorlxy — CHANGES_REQUESTED** [review]

**Comment by Mister-Hope** [comment]
Got it. i will update these days!

**Comment by ramesh-dada** [comment]
@meteorlxy @Mister-Hope THANK You so much for doing great work is their any estimated release date that when this plugin will be released ?

**Comment by Mister-Hope** [comment]
I am finishing all the changes, except I remain zh jsdocs in `SitemapOptions`.

IMO, as we are providing both chinese and english docs, it's reasonable to keep bothe languages in the final options provided to users to make sure they get full hints.

File: packages/@vuepress/plugin-sitemap/package.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "clean": "rimraf lib *.tsbuildinfo",
+    "copy": "cpx \"src/**/*.d.ts\" lib"
+  },
+  "dependencies": {
+    "@vuepress/client": "2.0.0-beta.21",
+    "@vuepress/core": "2.0.0-beta.21",
+    "@vuepress/plugin-git": "2.0.0-beta.21",
+    "@vuepress/shared": "2.0.0-beta.21",
+    "@vuepress/utils": "2.0.0-beta.21",
+    "chalk": "^4.1.1",
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/package.json:37]: `chalk` is also included in `@vuepress/utils`

File: packages/@vuepress/plugin-sitemap/package.json [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    "build": "tsc -b tsconfig.build.json",
+    "clean": "rimraf lib *.tsbuildinfo",
+    "copy": "cpx \"src/**/*.d.ts\" lib"
+  },
+  "dependencies": {
+    "@vuepress/client": "2.0.0-beta.21",
+    "@vuepress/core": "2.0.0-beta.21",
+    "@vuepress/plugin-git": "2.0.0-beta.21",
+    "@vuepress/shared": "2.0.0-beta.21",
+    "@vuepress/utils": "2.0.0-beta.21",
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/package.json:36]: Sync the version

File: packages/@vuepress/plugin-sitemap/src/node/types/options.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * 时间格式化器
+   *
+   * Date format function
+   */
+  dateFormatter?: DateFormatter
+
+  xslUrl?: string
+
+  /** XML namespaces to turn on - all by default */
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/src/node/types/options.ts:63]: Keep the comment style consistent

File: yarn.lock [hunk 10]
```diff
@@ -2459,6 +2459,13 @@
   dependencies:
     "@types/node" "*"
 
+"@types/fs-extra@^9.0.12":
+  version "9.0.12"
+  resolved "https://registry.yarnpkg.com/@types/fs-extra/-/fs-extra-9.0.12.tgz#9b8f27973df8a7a3920e8461517ebf8a7d4fdfaf"
+  integrity sha512-I+bsBr67CurCGnSenZZ7v94gd3tc3+Aj2taxMT4yu4ABLuOgOjeFxX3dokG24ztSRg5tnT00sL8BszO7gSMoIw==
+  dependencies:
+    "@types/node" "*"
+
```
Comment (meteorlxy) [inline yarn.lock:2468]: Should not be committed

File: docs/reference/plugin/sitemap.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+### sitemap.changefreq
+
+- Type: `'always'|'hourly'|'daily'|'weekly'|'monthly'|'yearly'|'never'`
+
+- Default: `'daily'`
+
+- Details:
+
+  Page default update frequency
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:17]: Missing ending period

File: docs/reference/plugin/sitemap.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+### sitemap.priority
+
+- Type: `number`
+
+- Default: `0.5`
+
+- Details:
+
+  Page priority, range from 0 - 1
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:37]:
```suggestion
  Page priority, range from `0` to `1`.
```

**Review by meteorlxy — CHANGES_REQUESTED** [review]

**Comment by meteorlxy** [comment]
> IMO, as we are providing both chinese and english docs, it's reasonable to keep both the languages in the final options provided to users to make sure they get full hints.

Remove them, or we need to add translations to all other comments.

**Comment by maodou38** [comment]
it's happy to know that we will support support sitemap recently.And also appreciate @Mister-Hope ‘s sitemap plugin.
i will use your plugin until the official release version online

**Comment by ramesh-dada** [comment]
@Mister-Hope Before you merge this can you please take a look at this also https://github.com/vuepress/vuepress-next/issues/353

**Comment by Mister-Hope** [comment]
I would prefer not to support any of the 2 fr in #353.  For my personal reason, see https://github.com/vuepress/vuepress-next/issues/353#issuecomment-898859697

If you have different ideas and think any of them should be supported, just leave a message.

File: packages/@vuepress/plugin-sitemap/src/node/types/sitemap.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
+export interface SitemapImageOption {
+  /** The URL of the image. */
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/src/node/types/sitemap.ts:2]: Normalize comment format

File: yarn.lock [hunk 14]
```diff
@@ -2598,6 +2598,11 @@
   resolved "https://registry.yarnpkg.com/@types/node/-/node-15.12.1.tgz#9b60797dee1895383a725f828a869c86c6caa5c2"
   integrity sha512-zyxJM8I1c9q5sRMtVF+zdd13Jt6RU4r4qfhTd7lQubyThvLfx6yYekWSQjGCGV2Tkecgxnlpl/DNlb6Hg+dmEw==
 
+"@types/node@^15.0.1":
+  version "15.14.2"
+  resolved "https://registry.yarnpkg.com/@types/node/-/node-15.14.2.tgz#7af8ab20156586f076f4760bc1b3c5ddfffd1ff2"
+  integrity sha512-dvMUE/m2LbXPwlvVuzCyslTEtQ2ZwuuFClDrOQ6mp2CenCg971719PTILZ4I6bTP27xfFFc+o7x2TkLuun/MPw==
+
```
Comment (meteorlxy) [inline yarn.lock:2605]: Run `npx yarn-deduplicate` to reduce duplicated lock records

File: docs/zh/reference/plugin/sitemap.md [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+### sitemap.changefreq
+
+- 类型： `'always'|'hourly'|'daily'|'weekly'|'monthly'|'yearly'|'never'`
+
+- 默认值： `'daily'`
+
+- 详情：
+
+  页面默认更新频率
```
Comment (meteorlxy) [inline docs/zh/reference/plugin/sitemap.md:17]: 缺少句号，下同

**Review by meteorlxy — CHANGES_REQUESTED** [review]

**Comment by maodou38** [comment]
why the pull request bog down？

**Comment by liziwl** [comment]
Any progress on the feature of sitemap?

**Comment by KnorpelSenf** [comment]
What are the chances that this lands in the 2.0 release?

**Comment by jrcharles** [comment]
Any update here?

**Comment by Mister-Hope** [comment]
I will finish it once my winter vacation begins, just busy being a postgraduate student studying quantum physics

**Comment by Mister-Hope** [comment]
@meteorlxy Should be ready

File: docs/reference/plugin/sitemap.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+```bash
+npm i -D @vuepress/plugin-sitemap@next
+```
+
+## Frontmatter Options
+
+### sitemap.changefreq
+
+- Type: `'always'|'hourly'|'daily'|'weekly'|'monthly'|'yearly'|'never'`
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:17]: Spaces around `|`

File: docs/reference/plugin/sitemap.md [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+Generate sitemap for your vuepress project.
+
+## Install
+
+```bash
+npm i -D @vuepress/plugin-sitemap@next
+```
+
+## Frontmatter Options
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:13]: Consistent with docs of other plugin.

`Frontmatter Options` -> `Frontmatter`

File: docs/reference/plugin/sitemap.md [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+- Type: `number`
+
+- Default: `0.5`
+
+- Details:
+
+  Page priority, range from `0` to `1`.
+
+## Plugin options
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:45]: Consistent with docs of other plugin.

`Plugin options` -> `Options`

File: docs/reference/plugin/sitemap.md [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+### outFile
+
+- Type: `string`
+
+- Default: `'sitemap.xml'`
+
+- Details:
+
+  The output filename.
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:85]: More description about the filename, is it relative to dest or somewhere?

File: docs/reference/plugin/sitemap.md [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+- Type: `string[]`
+
+- Default: `[]`
+
+- Details:
+
+  Pages not included.
+
+### outFile
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:77]: What about `outputFile`? Or using `sitemapFilename` like PWA plugin does?

File: docs/reference/plugin/sitemap.md [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+### sitemap.changefreq
+
+- Type: `'always'|'hourly'|'daily'|'weekly'|'monthly'|'yearly'|'never'`
+
+- Default: `'daily'`
+
+- Details:
+
+  Page default update frequency.
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:23]: More description about the differences between frontmatter and options, i.e. should specify that frontmatter would override options in current page.

Also, should add reference to corresponding options, and vice versa.

File: docs/reference/plugin/sitemap.md [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+### exclude
+
+- Type: `string[]`
+
+- Default: `[]`
+
+- Details:
+
+  Pages not included.
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:75]: Need more detailed explanation. 

- File path or route path?
- Relative path or absolute path?

File: docs/reference/plugin/sitemap.md [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+### dateFormatter
+
+- Type: `($page: Page) => string`
+
+- Required: false
+
+- Details:
+
+  Time formatter. The timestamp will be automatically generated through git plugin by default.
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:105]: Why a time / date formatter accepting a `page` arg ? Consider renaming it to `dateGetter` or so.

Also, it's better to provide an example.

File: docs/reference/plugin/sitemap.md [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+- Details:
+
+  Page default update frequency.
+
+### dateFormatter
+
+- Type: `($page: Page) => string`
+
+- Required: false
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:101]: Wrap false with backticks

File: packages/@vuepress/plugin-sitemap/tsconfig.build.json [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+  "compilerOptions": {
+    "module": "CommonJS",
+    "rootDir": "./src",
+    "outDir": "./lib"
+  },
+  "include": ["./src"],
+  "references": [
+    { "path": "../core/tsconfig.build.json" },
+    { "path": "../utils/tsconfig.build.json" }
+  ]
```
Comment (meteorlxy) [inline packages/@vuepress/plugin-sitemap/tsconfig.build.json:12]: plugin-git

**Review by meteorlxy — CHANGES_REQUESTED** [review]

**Comment by Mister-Hope** [comment]
Some explaination:

- A lot of plugin option has been renamed
- `modifyTimeGetter` is better with a `Page` arg, see plugin docs example
- A short description of sitemap is added.

File: docs/reference/plugin/sitemap.md [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+
+For use by search engines and other crawlers, there is a structured format, the XML Sitemap, which lists the pages in a site, their relative importance, and how often they are updated. This is pointed to from the robots.txt file and is typically called sitemap.xml.
+
+## Options
+
+### hostname
+
+- Type: `string`
+
+- Required: true
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:25]: backtick

File: docs/reference/plugin/sitemap.md [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+
+- Details:
+
+  The domain name where the current site is deployed, the plugin needs this option to work.
+
+### extraURLs
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:31]: Should be `extraUrls` ?

File: docs/reference/plugin/sitemap.md [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+
+- Details:
+
+  Extra links to be included.
+
+  If you have some links not including in vuepress router (normally in public folder or generated by other tools directly), you may need this option.
+
+  E.g.: `['/about.html', '/api/']`
+
+### excludePages
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:45]: Should be `excludeUrls`?

File: docs/reference/plugin/sitemap.md [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+
+- Example:
+
+  ```ts
+  // based on last modify time of source file
+  {
+    modifyTimeGetter: (page) =>
+      fs.statSync(app.dir.source(page.filePathRelative)).mtime.toISOString()
+  }
+  ```
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:95]: Remove the indents of code block

File: docs/reference/plugin/sitemap.md [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+
+```bash
+npm i -D @vuepress/plugin-sitemap@next
+```
+
+## Sitemap
+
+Sitemaps may be addressed to users or to software. Many sites have user-visible sitemaps which present a systematic view, typically hierarchical, of the site. These are intended to help visitors find specific pages, and can also be used by crawlers. Alphabetically organized site maps, sometimes called site indexes, are a different approach.
+
+For use by search engines and other crawlers, there is a structured format, the XML Sitemap, which lists the pages in a site, their relative importance, and how often they are updated. This is pointed to from the robots.txt file and is typically called sitemap.xml.
```
Comment (meteorlxy) [inline docs/reference/plugin/sitemap.md:17]: English and Chinese version is inconsistent 🤔

**Review by meteorlxy — CHANGES_REQUESTED** [review]

**Comment by Zhengqbbb** [comment]
@Mister-Hope Hi
Just used it today. The feeling is the same as imagined, very good feeling, but at the same time there are problems and bugs.

1. The `priority` configuration item in the document, (docs/zh/reference/plugin/sitemap.md: 85).Until I found that there is no SitemapOptions type, it means that **the configuration will not take effect**.This is very good design for normal pages.
2. About `robots.txt` generate options. see: https://developers.google.com/search/docs/advanced/robots/robots_txt
3. The `excludeUrls` default option not work. If I don't add the option manually, the 404 page will not be excluded
image
4. The `exclude`Frontmatter not work until I configured `excludeUrls` option.

**Comment by Mister-Hope** [comment]
> @Mister-Hope Hi Just used it today. The feeling is the same as imagined, very good feeling, but at the same time there are problems and bugs.
> 
> 1. The `priority` configuration item in the document, (docs/zh/reference/plugin/sitemap.md: 85).Until I found that there is no SitemapOptions type, it means that **the configuration will not take effect**.This is very good design for normal pages.
> 2. About `robots.txt` generate options. see: https://developers.google.com/search/docs/advanced/robots/robots_txt
> 3. The `excludeUrls` default option not work. If I don't add the option manually, the 404 page will not be excluded
>    image
> 4. The `exclude`Frontmatter not work until I configured `excludeUrls` option.

Hi, thanks for the feed back, I will have a look later on 1,3,4 , but I am not catching what you mean about 2. All the output folder is deployed directly, so if user set a base, there is no way for me to generate a robot.txt and place outside the dest folder. The robots.txt should be ignored by google. Also, I am not sure all the search engine won't read subfolder robot.txt. Besides, a sitemap plugin should not try to set `allow` and `disallow` for developers, in my case, I am having a valid robot.txt in my project public folder, and I only want the plugin appends sitemap link for me. If you have any idea (which is better), please point it out since I have no idea how to improve it now.

**Comment by Zhengqbbb** [comment]
> > @Mister-Hope 
> Hi, thanks for the feed back, I will have a look later on 1,3,4 , but I am not catching what you mean about 2. All the output folder is deployed directly, so if user set a base, there is no way for me to generate a robot.txt and place outside the dest folder. The robots.txt should be ignored by google. Also, I am not sure all the search engine won't read subfolder robot.txt. Besides, a sitemap plugin should not try to set `allow` and `disallow` for developers, in my case, I am having a valid robot.txt in my project public folder, and I only want the plugin appends sitemap link for me. If you have any idea (which is better), please point it out since I have no idea how to improve it now.

ok, I just think that if there is automatic generate robots.txt, it is better to have a better place to configure.
Maybe someone needs to disable the search spider to crawl their own website to prevent it from being website included, such as only for Google Spider, we can provide option to set user-agent.
If you want to uniformly set recommended website inclusion and non-recommended website inclusion, you need Allow and Disallow.

**Comment by Mister-Hope** [comment]
I will change the logic, only when base is `/` and user has a robot.txt in public folder, the plugin will try to add sitemap url to it. This should be better.

**Comment by Mister-Hope** [comment]
And for 1, you should set priority in `frontmatter.sitemap.priority` and it's injected in the speard operator I think.

```
    const sitemapInfo: SitemapPageInfo = {
      changefreq,
      links,
      ...(lastmodifyTime ? { lastmod: lastmodifyTime } : {}),
      ...frontmatterOptions,
    }
```

Is this option problematic to you?

**Comment by Mister-Hope** [comment]
The missing line in https://github.com/vuepress/vuepress-next/pull/277/commits/f043211d2cea84c83396cf1f91c1123fa4b7d22a should solve both 3 and 4 @Zhengqbbb Thanks for point out the bug.

**Comment by Zhengqbbb** [comment]
> I will change the logic, only when base is `/` and user has a robot.txt in public folder, the plugin will try to add sitemap url to it. This should be better.

I also think this is the right design, **whatever this is a sitemap plugin**. 
But robot.txt is spider friendly for sites that don't submit sitemaps. **For documentation**, it would be better to mention that robot.txt needs to be added to the publish folder.

**Comment by Zhengqbbb** [comment]
> And for 1, you should set priority in `frontmatter.sitemap.priority` and it's injected in the speard operator I think.
> Is this option problematic to you?

- But you can see that **it is not mentioned in the English document**, but mentioned in the one asked in Chinese.
- I think the significance of this option item is that for `md ` files that do not declare formatter on normal pages, this is a **unified** source of priority. After all, everyone is not very willing to think about the priority of this page.

**Comment by Zhengqbbb** [comment]
> > And for 1, you should set priority in `frontmatter.sitemap.priority` and it's injected in the speard operator I think.
> > Is this option problematic to you?
> 
> * But you can see that **it is not mentioned in the English document**, but mentioned in Chinese document.
> * I think the significance of this option item is that for `md ` files that do not declare formatter on normal pages, this is a **unified** source of priority. After all, everyone is not very willing to think about the priority of this page.

**Comment by Mister-Hope** [comment]
> 

The missing option in docs is added.

**Comment by Mister-Hope** [comment]
> > And for 1, you should set priority in `frontmatter.sitemap.priority` and it's injected in the speard operator I think.
> > Is this option problematic to you?
> 
> * But you can see that **it is not mentioned in the English document**, but mentioned in the one asked in Chinese.
> * I think the significance of this option item is that for `md ` files that do not declare formatter on normal pages, this is a **unified** source of priority. After all, everyone is not very willing to think about the priority of this page.

A description is added

**Comment by Mister-Hope** [comment]
@meteorlxy Could you have another check?

**Comment by Jelledb** [comment]
Looking forward to the sitemap plugin!

**Comment by JohannesRudolph** [comment]
I can confirm the sitemap plugin is working fine, pulled the code from the PR in as a "vendored" dependency and it works great for our site.

**Comment by JohannesRudolph** [comment]
One feedback, I'd like to have the option to set the `changefreq` option to null/undefined so that it's not part of the sitemap. Since we already have `lastmod` this is going to be fine for most sites and sends less confusing signals to search engine crawlers.

**Comment by hustcer** [comment]
It's really a long time to wait

**Comment by Mister-Hope** [comment]
You can use `vuepress-plugin-sitemap2`

**Comment by Zhengqbbb** [comment]
> Any updates?
> 
> Also, I found this site helpful:
> 
> https://developers.google.com/search/docs/advanced/crawling/localized-versions#sitemap

Hey u can use `vuepress-plugin-sitemap2`
Demo:
https://github.com/Zhengqbbb/vuepress-plugin/blob/b5383fa7d8a548d8306f5ab49e60a4a567281874/docs/.vuepress/config.ts#L42-L45

**Comment by Mister-Hope** [comment]
That plugin is published by me as the same with this pr.

**Comment by taozuhong** [comment]
能不能早点合入并发布稳定版本？要不 VitePress 都出来了，它的价值空间就会被严重挤压，这么多努力就白费了。

**Comment by meteorlxy** [comment]
@taozuhong You can take a look at https://github.com/vuejs/vitepress/discussions/548

VitePress might be VuePress 3 as a decision of Vue.js Team. So VuePress 2 has already been in an awkward position. That's why the stable version delayed a lot.

**Comment by Mister-Hope** [comment]
I am updating `vuepress-plugin-sitemap2` with every vp2 version, you are free to use that.

**Comment by taozuhong** [comment]
> @taozuhong You can take a look at [vuejs/vitepress#548](https://github.com/vuejs/vitepress/discussions/548)
> 
> VitePress might be VuePress 3 as a decision of Vue.js Team. So VuePress 2 has already been in an awkward position. That's why the stable version delayed a lot.

Nice to see it, hope the plugins Api become more stable than last, make the plugin ecosystem release more power.

**Comment by Mister-Hope** [comment]
Closing, this will be maintained speratedly

**Pull request closed without merging**